### PR TITLE
fix(weaviate): guard empty objectIds in mutation operations

### LINKED ISSUE #348: Weaviate: empty objectIds can trigger mass delete / mass state mutation
### Description

`nodes/src/nodes/weaviate/weaviate.py` mutating methods (`remove`, `markDeleted`, `markActive`) do not guard empty `objectIds` input.

When `objectIds=[]`, the code still builds `Filter.any_of([])` and executes delete/update queries. In current behavior this can broaden scope and apply to all records in the collection.

This is a **data-integrity risk** with potential large blast radius (mass delete or mass `isDeleted` toggle), so severity is treated as **P0**.

### Steps to Reproduce

1. Check out current `develop`.
2. Run the following deterministic repro script:

```bash
cd /Users/krishgarg/Documents/Projects/RocketRide/rocketride-server
python3 - <<'PY'
import importlib.util, sys, types
from pathlib import Path

root = Path('.').resolve()
sys.path.insert(0, str(root / 'nodes' / 'test' / 'mocks'))

mod_depends = types.ModuleType('depends')
mod_depends.depends = lambda *_args, **_kwargs: None
sys.modules['depends'] = mod_depends

ai_pkg = types.ModuleType('ai')
common_pkg = types.ModuleType('ai.common'); common_pkg.__path__ = []
schema_mod = types.ModuleType('ai.common.schema')
store_mod = types.ModuleType('ai.common.store')
config_mod = types.ModuleType('ai.common.config')

class Doc: pass
class DocFilter: pass
class DocMetadata: pass
class QuestionText: pass
class DocumentStoreBase:
    def __init__(self, *_args, **_kwargs):
        pass
class Config:
    @staticmethod
    def getNodeConfig(_provider, _connConfig):
        return {}

schema_mod.Doc = Doc
schema_mod.DocFilter = DocFilter
schema_mod.DocMetadata = DocMetadata
schema_mod.QuestionText = QuestionText
store_mod.DocumentStoreBase = DocumentStoreBase
config_mod.Config = Config

sys.modules['ai'] = ai_pkg
sys.modules['ai.common'] = common_pkg
sys.modules['ai.common.schema'] = schema_mod
sys.modules['ai.common.store'] = store_mod
sys.modules['ai.common.config'] = config_mod

weaviate_file = root / 'nodes' / 'src' / 'nodes' / 'weaviate' / 'weaviate.py'
spec = importlib.util.spec_from_file_location('weaviate_store_under_test', weaviate_file)
module = importlib.util.module_from_spec(spec)
assert spec and spec.loader
spec.loader.exec_module(module)
Store = module.Store

import weaviate

collection = weaviate.MockCollection('test-coll')
collection._storage.clear()
collection._storage['r1'] = {'properties': {'objectId': 'obj-1', 'isDeleted': False}, 'vector': [0.1, 0.2]}
collection._storage['r2'] = {'properties': {'objectId': 'obj-2', 'isDeleted': False}, 'vector': [0.3, 0.4]}

store = Store.__new__(Store)
store.collectionObj = collection
store.doesCollectionExist = lambda *_args, **_kwargs: True

print('before remove([]):', sorted(collection._storage.keys()))
store.remove([])
print('after remove([]):', sorted(collection._storage.keys()))

collection._storage['r1'] = {'properties': {'objectId': 'obj-1', 'isDeleted': False}, 'vector': [0.1, 0.2]}
collection._storage['r2'] = {'properties': {'objectId': 'obj-2', 'isDeleted': False}, 'vector': [0.3, 0.4]}
store.markDeleted([])
print('after markDeleted([]):', [collection._storage[k]['properties']['isDeleted'] for k in sorted(collection._storage.keys())])
PY
```

### Expected vs Actual Behavior

**Expected**
- Empty `objectIds` should be treated as no-op for `remove`, `markDeleted`, and `markActive`.
- No records should be mutated/deleted.

**Actual**
- `remove([])` deletes all records in the collection.
- `markDeleted([])` marks all records as deleted.

### Severity

P0 - Critical (service down, data loss)

### Platform

- [x] Windows
- [x] macOS
- [x] Linux

### Version / Environment

`develop` branch (as of 2026-03-22), local deterministic repro with repository mocks.

### Screenshots / Logs

```text
before remove([]): ['r1', 'r2']
after remove([]): []
after markDeleted([]): [True, True]
```

### PR DESCRIPTION
**Pull request by kgarg2468**
## Summary
- Prevent broad Weaviate mutations when `objectIds` is empty by making `remove`, `markDeleted`, and `markActive` explicit no-ops.
- Add regression tests that reproduce the pre-fix mass-mutation behavior and verify both empty-list no-op and normal targeted behavior.

## Bug
In `nodes/src/nodes/weaviate/weaviate.py`, mutation methods build `Filter.any_of(objectIdsFilter)` without guarding empty input.

When `objectIds=[]`, `Filter.any_of([])` can act as an unscoped match in current behavior, causing operations intended for specific object IDs to affect all records.

## Steps to Reproduce
```bash
cd /Users/krishgarg/Documents/Projects/RocketRide/rocketride-server
python3 - <<'PY'
import importlib.util, sys, types
from pathlib import Path

root = Path('.').resolve()
sys.path.insert(0, str(root / 'nodes' / 'test' / 'mocks'))

mod_depends = types.ModuleType('depends')
mod_depends.depends = lambda *_args, **_kwargs: None
sys.modules['depends'] = mod_depends

ai_pkg = types.ModuleType('ai')
common_pkg = types.ModuleType('ai.common'); common_pkg.__path__ = []
schema_mod = types.ModuleType('ai.common.schema')
store_mod = types.ModuleType('ai.common.store')
config_mod = types.ModuleType('ai.common.config')

class Doc: pass
class DocFilter: pass
class DocMetadata: pass
class QuestionText: pass
class DocumentStoreBase:
    def __init__(self, *_args, **_kwargs):
        pass
class Config:
    @staticmethod
    def getNodeConfig(_provider, _connConfig):
        return {}

schema_mod.Doc = Doc
schema_mod.DocFilter = DocFilter
schema_mod.DocMetadata = DocMetadata
schema_mod.QuestionText = QuestionText
store_mod.DocumentStoreBase = DocumentStoreBase
config_mod.Config = Config

sys.modules['ai'] = ai_pkg
sys.modules['ai.common'] = common_pkg
sys.modules['ai.common.schema'] = schema_mod
sys.modules['ai.common.store'] = store_mod
sys.modules['ai.common.config'] = config_mod

weaviate_file = root / 'nodes' / 'src' / 'nodes' / 'weaviate' / 'weaviate.py'
spec = importlib.util.spec_from_file_location('weaviate_store_under_test', weaviate_file)
module = importlib.util.module_from_spec(spec)
assert spec and spec.loader
spec.loader.exec_module(module)
Store = module.Store

import weaviate

collection = weaviate.MockCollection('test-coll')
collection._storage.clear()
collection._storage['r1'] = {'properties': {'objectId': 'obj-1', 'isDeleted': False}, 'vector': [0.1, 0.2]}
collection._storage['r2'] = {'properties': {'objectId': 'obj-2', 'isDeleted': False}, 'vector': [0.3, 0.4]}

store = Store.__new__(Store)
store.collectionObj = collection
store.doesCollectionExist = lambda *_args, **_kwargs: True

print('before remove([]):', sorted(collection._storage.keys()))
store.remove([])
print('after remove([]):', sorted(collection._storage.keys()))

collection._storage['r1'] = {'properties': {'objectId': 'obj-1', 'isDeleted': False}, 'vector': [0.1, 0.2]}
collection._storage['r2'] = {'properties': {'objectId': 'obj-2', 'isDeleted': False}, 'vector': [0.3, 0.4]}
store.markDeleted([])
print('after markDeleted([]):', [collection._storage[k]['properties']['isDeleted'] for k in sorted(collection._storage.keys())])
PY
```

## Root Cause
`remove`, `markDeleted`, and `markActive` did not short-circuit on empty `objectIds`. They still built an OR filter from an empty condition list and executed mutations with that filter.

## Fix
- Added `if not objectIds: return` to:
  - `Store.remove`
  - `Store.markDeleted`
  - `Store.markActive`
- Added regression tests in `nodes/test/test_weaviate_record_updates.py`:
  - empty `objectIds` is no-op for all three methods
  - non-empty `objectIds` still mutates only matching records

## Why This Works
The methods now enforce explicit input semantics:
- empty target set => no-op
- non-empty target set => filtered mutation only

This removes the implicit reliance on backend filter behavior for empty OR clauses and prevents unintended collection-wide mutation.

## Testing / Validation
- Added and ran targeted regression tests:
  - `cd /Users/krishgarg/Documents/Projects/RocketRide/rocketride-server`
  - `.venv/bin/pytest nodes/test/test_weaviate_record_updates.py -v`
  - Result: `6 passed`
- Attempted requested broad test command:
  - `./builder test --verbose`
  - Result: failed before tests due local toolchain prerequisite: missing CMake (`Checking build prerequisites... CMake is not installed`)
- Attempted `nodes/test/test_pinecone_record_updates.py` per prior plan, but that file does not exist on current `develop`.

## Type
fix

## Checklist
- [x] Tests added/updated
- [x] Tested locally
- [ ] `./builder test` passes (blocked locally by missing CMake prerequisite in builder bootstrap)
- [x] Commit messages follow [conventional commits](https://www.conventionalcommits.org/)
- [x] No secrets or credentials included

## Related Issues
Closes #348

## Summary by CodeRabbit

* **Improvements**
  * Enhanced performance of record deletion and update operations with optimized handling of empty inputs

* **Tests**
  * Added regression tests verifying Weaviate record deletion and update operations correctly handle empty input and only affect targeted records

#frontier-tower-hackathon

## Impact

This prevents unintended broad mutations when callers pass an empty ID list, protecting existing records from accidental updates.

## CI Evidence

- [GitHub PR checks for #349](https://github.com/rocketride-org/rocketride-server/pull/349/checks)

### CODE REVIEW

File: nodes/test/test_weaviate_record_updates.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+
+def _make_obj_records(object_id: str, count: int, is_deleted: bool) -> dict[str, dict]:
+    return {
+        f'{object_id}-{index}': {
+            'properties': {'objectId': object_id, 'isDeleted': is_deleted},
+            'vector': [0.1, 0.2],
+        }
+        for index in range(count)
+    }
```
[the remark this answers is not in the document]
Comment (kgarg2468) [reply nodes/test/test_weaviate_record_updates.py]: Updated `_make_obj_records` to use a keyword-only boolean parameter (`*, is_deleted: bool`). Existing call sites already pass `is_deleted=` explicitly. Re-ran `nodes/test/test_weaviate_record_updates.py` and all 6 tests pass.

File: nodes/test/weaviate/test_record_updates.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+
+def _make_obj_records(object_id: str, count: int, *, is_deleted: bool) -> dict[str, dict]:
+    return {
+        f'{object_id}-{index}': {
+            'properties': {'objectId': object_id, 'isDeleted': is_deleted},
+            'vector': [0.1, 0.2],
+        }
+        for index in range(count)
+    }
```
[the remark this answers is not in the document]
Comment (kgarg2468) [reply nodes/test/weaviate/test_record_updates.py:164]: Added a PEP 257 docstring to `_make_obj_records` describing purpose, parameters, and return shape. Kept the keyword-only signature unchanged and verified call sites still use `is_deleted=`. Re-ran `.venv/bin/pytest nodes/test/test_weaviate_record_updates.py -v` with 6/6 tests passing.

File: nodes/test/test_weaviate_record_updates.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    records = _make_obj_records('obj-1', 2, is_deleted=True)
+    records.update(_make_obj_records('obj-2', 1, is_deleted=True))
+    store, collection = _make_store(records)
+
+    store.markActive(['obj-1'])
+
+    obj1 = [record['properties']['isDeleted'] for record in collection._storage.values() if record['properties']['objectId'] == 'obj-1']
+    obj2 = [record['properties']['isDeleted'] for record in collection._storage.values() if record['properties']['objectId'] == 'obj-2']
+    assert obj1 == [False, False]
+    assert obj2 == [True]
```
[the remark this answers is not in the document]
Comment (kgarg2468) [reply nodes/test/test_weaviate_record_updates.py]: Refactored this section into two  tests (one for empty- no-op behavior and one for scoped mutations) while keeping the same behavioral assertions. Re-ran ============================= test session starts ==============================
platform darwin -- Python 3.13.5, pytest-9.0.2, pluggy-1.6.0 -- /Users/krishgarg/Documents/Projects/RocketRide/rocketride-server/.venv/bin/python
cachedir: .pytest_cache
rootdir: /Users/krishgarg/Documents/Projects/RocketRide/rocketride-server
configfile: pyproject.toml
plugins: asyncio-1.3.0
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collecting ... collected 6 items

nodes/test/test_weaviate_record_updates.py::test_mutation_noop_for_empty_objectids[remove-False] PASSED [ 16%]
nodes/test/test_weaviate_record_updates.py::test_mutation_noop_for_empty_objectids[markDeleted-False] PASSED [ 33%]
nodes/test/test_weaviate_record_updates.py::test_mutation_noop_for_empty_objectids[markActive-True] PASSED [ 50%]
nodes/test/test_weaviate_record_updates.py::test_mutation_updates_only_matching_object_ids[remove-False-expected_obj1_is_deleted0-expected_obj2_is_deleted0-expected_remaining_object_ids0] PASSED [ 66%]
nodes/test/test_weaviate_record_updates.py::test_mutation_updates_only_matching_object_ids[markDeleted-False-expected_obj1_is_deleted1-expected_obj2_is_deleted1-expected_remaining_object_ids1] PASSED [ 83%]
nodes/test/test_weaviate_record_updates.py::test_mutation_updates_only_matching_object_ids[markActive-True-expected_obj1_is_deleted2-expected_obj2_is_deleted2-expected_remaining_object_ids2] PASSED [100%]

============================== 6 passed in 0.05s =============================== and all 6 cases pass.

[the remark this answers is not in the document]
Comment (kgarg2468) [reply nodes/test/test_weaviate_record_updates.py]: Applied in commit d9b4b3d: refactored these into two parameterized tests (empty objectIds no-op + scoped mutation behavior) with equivalent assertions. Re-ran .venv/bin/pytest nodes/test/test_weaviate_record_updates.py -v and all 6 cases pass.
